### ISSUE TITLE
Singletons - correct syntax.

### ISSUE THREAD

**Issue by kostiakoval** [agensdev/playbook-ios#39]
How to make and declare a singleton. (if you thought twice and youReally, REALLY need it :D) 

```objc
+ (MyClass *)sharedInstance {
    static MyClass *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[MyClass alloc] init];
    });
    return sharedInstance;
}
```
👍 

vs

```objc
+ (MyClass *)sharedInstance {
    static MyClass *sharedInstance;
    @synchronized(self) {
        if (sharedInstance == nil) {
            sharedInstance = [[MyClass alloc] init];
        }
    }
    return sharedInstance;
}
```
❤️

**Comment by hfossli** [comment]
Depends on wether you need to nil out the singleton in unit tests or not

**Comment by kostiakoval** [comment]
if you need to nil clean your singleton, them maybe it shouldn't be a singleton in the first place 😉 🤔 
Or maybe the component that uses that singleton should have been passed it as a dependency. 

@hfossli @aarsland

**Comment by onmyway133** [comment]
I usually do the first one in ObjectiveC, but I use `instancetype` instead of `MyClass`

**Comment by kostiakoval** [comment]
How would it worked if you make a subclass of a singleton, then `+ (instancetype *)sharedInstance` would expect that return type `instancetype ` is a subclass, but it's not. 
Am I right? @onmyway133 ?

**Comment by onmyway133** [comment]
@kostiakoval If I used singleton, I would never subclass it. Maybe it's just me

**Comment by kostiakoval** [comment]
I think I never done that to, but then `MyClass` makes more sense, maybe?

**Comment by hfossli** [comment]
> Or maybe the component that uses that singleton should have receive it as a dependency
I agree, but that is not as trivial to do in objc (without causing a mess) as in swift.
